Start release notes for v0.2.0 (#68)
confidential-containers/documentation: releases/v0.2.0.md

# Release Notes for v0.2.0
Confidential Containers has adopted a six-week release cadence.
This is our first release on this schedule.
This release mainly features incremental improvements to our build system and tests
as well as minor features, adjustments, and cleanup.

Please see the [quickstart guide](../quickstart.md) for details on how to try out Confidential Containers

## What's new

- Kata CI uses existing Kata tooling to build components
- Kata CI caches build environments for components

> - Incremental advances in signature support

Author: I wasn't really sure what the status is. I'd be happy for someone in the know to add a separate entry for it.

> - Incremental advances in signature support

Reviewer: @Xynnn007 - do you have anything, you think we should add here on cosign? Thanks

> - Incremental advances in signature support

Reviewer 2: Yea, thanks, now 
```
- Support cosign-signed image verification
```

> - Incremental advances in signature support

Author: Ok, since it sounds like there are still some limitations I added the cosign support to the existing bullet about signature improvements. It will be great to have more documentation about cosign and signatures in general. I don't think many people know how to use it yet.

- KBS no longer required to use unencrypted images with SEV.

## Hardware Support
Confidential Containers is tested with attestation on the following platforms:
- Intel TDX
- AMD SEV

The following platforms are untested or partially supported:
- AMD SEV-ES
- IBM Z SE

The following platforms are in development:

> - Intel SGX

Reviewer 3: We currently have a test for Intel SGX about to be merged on the Operator.
So, I think this should be "bumped" to, at least" a new bullet saying

```
Confidential Containers is tested without attestation  attestation on the following platforms:
- Intel SGX
```

/cc @dcmiddle @mythi 
```

Reviewer: I assume - removing the duplicate ` attestation `?

Author: How about moving to it to the untested or partially supported category?

Reviewer 3: I'm not the one running the Enclave CC show, but I think it can be consiudered supported and tested.

We have a test about to be merged on the Operator side, it's just lacking the attestation part.

Author: I understand that talking about exactly what is supported has the potential to be dramatic.

My goal with the sections is really just to have 1) totally supported and has pretty solid tests for most features; basically done except for new features that are added to the project as a whole 2) almost 1 3) we're working on it. My impression is that `enclave-cc` goes in the second category although there has been a lot of great progress lately. We can adjust the names of the sections if that would make things more clear. 

wdyt @dcmiddle @mythi

Reviewer 4: Dramatic?!?! WHO IS GOING TO BE DRAMATIC?!?!?!?!?!
;-)
I'm actually fully onboard with listing SGX in level 2. We've got operator and some CI. We need to get more SGX-HW testing integrated. On the upside, though, @mythi got simulation-mode going which makes it easy to test without real SGX hardware. That way people can try it out on whatever platform they have access to.

Author: I was planning on being dramatic but maybe not.

- AMD SEV-SNP

## Limitations

The following are known limitations of this release:

- Platform support is currently limited, and rapidly changing
  * S390x is not supported by the CoCo operator
  * AMD SEV-ES has not been tested.
  * AMD SEV does not support container image signature validation.
- Attestation and key brokering support is still under development

>   * The disk-based key broker client (KBC) is used when there is no HW support, but is not suitable for production (except with encrypted VM images).

Reviewer 5: ```suggestion
  * The disk-based key broker client (KBC) is used when there is no confidential computing HW support, but is not suitable for production (except with encrypted VM images).
```

  * Currently, there are two KBS that can be used:
    - simple-kbs:  simple key broker service (KBS) for SEV(-ES).
    - [Verdictd](https://github.com/inclavare-containers/verdictd): An external project with which Attestation Agent can conduct remote attestation communication and key acquisition via EAA KBC
  * The full-featured generic KBS and the corresponding KBC are still in the development stage.
  * For developers, other KBCs can be experimented with.
  * AMD SEV must use a KBS even for unencrypted images.
- The format of encrypted container images is still subject to change
  * The oci-crypt container image format itself may still change
  * The tools to generate images are not in their final form
  * The image format itself is subject to change in upcoming releases

>   * Image repository support for encrypted images is unequal

Reviewer 2: And now due to https://github.com/kata-containers/kata-containers/issues/5582, now `S390x is not support by cosign-signed image verification`

Reviewer: @Xynnn007 - hey Ding, the situation is worse than this I'm afraid - due to the incompatibility with the `ring` crate and `s390x` we aren't able to build the kata-agent for `s390x` at all, so if we want to add co-sign support this release (which I'm not against), and document the restriction in the release notes, we'd just need to say that `s390x` isn't working at all for this release.

Reviewer 2: I got that. So due to this issue (incompatibility with the `ring` crate and `s390x`), what should we do now? I'm not sure, are there two different ways?
- Say that this release do not support `s390x` at all (for we can not build kata-agent now)
- Put a fix in `image-rs` to add a switch `cosign` verification (feature in `Cargo.toml`), and add compile configuration in `kata-agent` to disable cosign when the arch is `s390x`. Also, some CI fixes when the arch is `s390x`.

Reviewer 2: Or we can not add `cosign` this release? CC @peterzcst @arronwy

Reviewer: I think that due to `s390x` not being supported, the community will vote for
> Say that this release do not support s390x at all (for we can not build kata-agent now)

It is a bit unfortunately timing which that due to delayed backporting of an `s390x` from `main` into `CCv0` we missed spotting the failure in the PR that bumped image-rs, but as long as the community continue to try and help fix this for 0.3 (which I'll recognise you have been doing so far 🙇), then I won't suggest holding back this feature from the release.

A lesson for the future is that we need to ensure that we get some s390x tests as part of the image-rs CI like we've done for the attestation-agent. That's been on my list for a while though and is IBM's responsibility to drive.

Reviewer: Talking to @hbrueckner - we've been discussing the idea of pulling in an arch specific version of `image-rs` into the kata-agent. Then we could pull the `0.2` tag for non-s390x arch and the commit before the `ring` dependency got pulled in for s390x therefore losing co-sign support for s390x, but enabling compilation.

I don't like splitting the code like this and it would obviously not be supported, but is an idea to avoid a s390x regression for 0.2?

Reviewer 3: I'm fine with this approach.

Reviewer: I think that now we have https://github.com/kata-containers/kata-containers/pull/5600 merged, we might be back to Ding's original suggest of documenting the s390x limitation, so something like:
> Upstream Kata Agent cannot build for `s390x`, meaning that `s390x` misses some newer features, such as cosign support.

Reviewer 6: Oh well, (shamelessly stolen from the wrong place): with https://github.com/kata-containers/kata-containers/pull/5600 merged, kata-agents builds again. So it should more read like:

> s390x does not support container image signature with cosign in this release.

Reviewer: Thanks for the clarification!

- CoCo currently requires a custom build of `containerd`
  * The CoCo operator will deploy the correct version of `containerd` for you
  * Changes are required to delegate `PullImage` to the agent in the virtual machine
  * The required changes are not part of the vanilla `containerd`
  * The final form of the required changes in `containerd` is expected to be different
  * `crio` is not supported
- CoCo is not fully integrated with the orchestration ecosystem (Kubernetes, OpenShift)

>   * OpenShift is a non-started at the moment due to their dependency on CRIO

Reviewer 7: nit: non-started -> non-starter 
this typo is a leftover from the previous release so if possible better to fix the typo in this release :-)

>   * OpenShift is a non-started at the moment due to their dependency on CRIO

Reviewer 5: ```suggestion
  * OpenShift is a non-starter at the moment due to its dependency on [CRI-O](https://github.com/cri-o/cri-o)
```

Is Openshift really not supporting containerd?

Reviewer 3: Yep, it's tied to CRI-O.

  * Existing APIs do not fully support the CoCo security and threat model
  * Some commands accessing confidential data, such as `kubectl exec`, may either fail to work, or incorrectly expose information to the host
  * Container image sharing is not possible in this release

>   * Container images are downloaded by the guest (with encryption), not  by the host

Author: Well if we can't fix any of the limitations at least we can remove an extra space in the limitations.

  * As a result, the same image will be downloaded separately by every pod using it, not shared between pods on the same host.
- The CoCo community aspires to adopting open source security best practices, but not all practices are adopted yet.

>   * We track our status with the OpenSSF Best Practices Badge, which was at 43% at the time of this release.

Reviewer 4: ```suggestion
  * We track our status with the OpenSSF Best Practices Badge, which increased to 46% at the time of this release.
```

# Release Notes for v0.2.0
Confidential Containers has adopted a six-week release cadence.
This is our first release on this schedule.
This release mainly features incremental improvements to our build system and tests
as well as minor features, adjustments, and cleanup.

Please see the [quickstart guide](../quickstart.md) for details on how to try out Confidential Containers

## What's new

- Kata CI uses existing Kata tooling to build components.
- Kata CI caches build environments for components.
- Pod VM can be launched with measured boot. See [more info](https://github.com/confidential-containers/documentation/issues/40)
- Incremental advances in signature support including verification of cosign-signed images.

> - KBS no longer required to use unencrypted images with SEV.

Reviewer 4: ```suggestion
- KBS no longer required to use unencrypted images with SEV.
- Enclave-cc added to operator providing initial SGX support. 
```

- More rigorous versioning of sub-projects

## Hardware Support
Confidential Containers is tested with attestation on the following platforms:
- Intel TDX
- AMD SEV

The following platforms are untested or partially supported:
- Intel SGX
- AMD SEV-ES
- IBM Z SE

The following platforms are in development:
- AMD SEV-SNP

## Limitations

The following are known limitations of this release:

- Platform support is currently limited, and rapidly changing

>   * S390x is not supported by the CoCo operator

Reviewer 6: @fitzthum

I like to have the `s390x` vs. `s390` vs. IBM Z SE etc. being used more consistently. However, I see the rush for get the v0.2.0 release out.  Just an idea to discuss: It would be nice to create the release notes for the v0.3.0 right after the v0.2.0 is out, then we can go even with feature PRs updating the release notes as we go along the timeline. Wdyt?

Author: That's a good idea. I will try to remember to do this.

Author: Also let me at least make that S lowercase. I try to mention Z at the top as the Confidential Computing feature that is supported but then the technical stuff is supposed to say `s390x` to talk about the OS. These release notes were written by a few different people, though.

  * The full-featured generic KBS and the corresponding KBC are still in the development stage.
  * For developers, other KBCs can be experimented with.
  * AMD SEV must use a KBS even for unencrypted images.
- The format of encrypted container images is still subject to change
  * The oci-crypt container image format itself may still change
  * The tools to generate images are not in their final form
  * The image format itself is subject to change in upcoming releases
  * Image repository support for encrypted images is unequal
- CoCo currently requires a custom build of `containerd`
  * The CoCo operator will deploy the correct version of `containerd` for you
  * Changes are required to delegate `PullImage` to the agent in the virtual machine
  * The required changes are not part of the vanilla `containerd`
  * The final form of the required changes in `containerd` is expected to be different
  * `crio` is not supported
- CoCo is not fully integrated with the orchestration ecosystem (Kubernetes, OpenShift)

>   * OpenShift is a non-starter at the moment due to their dependency on CRIO

Reviewer 5: ```suggestion
  * * OpenShift is a non-starter at the moment due to its dependency on [CRI-O](https://github.com/cri-o/cri-o)
```

>   * Existing APIs do not fully support the CoCo security and threat model

Reviewer 8: ```suggestion
Existing APIs do not [fully support]( https://docs.google.com/presentation/d/16649JpQAdDb3jh3OVKWdNks0mvcqpidssrm0pOZNKZ4/edit#slide=id.p) the CoCo security and threat model

  * Some commands accessing confidential data, such as `kubectl exec`, may either fail to work, or incorrectly expose information to the host
  * Container image sharing is not possible in this release
  * Container images are downloaded by the guest (with encryption), not by the host

>   * As a result, the same image will be downloaded separately by every pod using it, not shared between pods on the same host.

Reviewer 8: This is a new copy of the slide deck, which should be public. Please verify that you have access.

>   * As a result, the same image will be downloaded separately by every pod using it, not shared between pods on the same host.

Author: I can access. I'm going to add a link to the issue about image pull instead. Let's link these slides from that issue.

>   * As a result, the same image will be downloaded separately by every pod using it, not shared between pods on the same host.

Reviewer 8: A better approach, @fitzthum. Thanks.
